### PR TITLE
Added unit_tests.py 

### PR DESCRIPTION
**Pull request by styoung3**
Basic testing framework for functions in scaffolding.py, specifically energy_intensity and energy_impact calculations. Other things were added that need to be cleaned up at this point, such as time and cost impact functions which are just duplicates of energy_intensity and energy_impact calculations with a few modifications. Also, the notebook outputs need to be cleared.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -133,3 +133,10 @@ dmypy.json
 
 # Pyre type checker
 .pyre/
+<<<<<<< Updated upstream
+.DS_Store
+.DS_Store
+viz_scripts/.DS_Store
+=======
+viz_scripts/auxiliary_files/.DS_Store
+>>>>>>> Stashed changes
```
Comment (shankari) [inline .gitignore:142]: This is a merge conflict. You need to resolve it properly before committing

File: OpenSourceLicenses.md [hunk 2]
```diff
@@ -9,5 +9,7 @@ The binary distributions are directly from the `gh-pages` branch.
 
 The bootstrap template is from:
 https://github.com/StartBootstrap/startbootstrap-bare.git
-with a modified contents, modified navbar
-and theming based on the original web page
+which is also available under an MIT license.
+
+This repo includes modified content, modified navbar
+and theming based on the original web page.
```
Comment (shankari) [inline OpenSourceLicenses.md:15]: This is a change that I made to main 9 days ago.
https://github.com/e-mission/em-public-dashboard/commit/893fd0a3a5e6a9ccded82d1793459a25450d30f8

Not sure why it is showing up as a change to your branch. Did you not pull from main before making your change?

**Review by shankari — COMMENTED** [review]
Please remove outputs from `.ipynb` files by opening the notebook and then using Kernel -> Restart and clear outputs.

File: .gitignore (hunk 1 above)
Comment (shankari) [reply .gitignore:145]: @styoung3 the reason that you see `.DS_Store` over and over again is that your `.gitignore` is broken due to merge conflicts.

File: viz_scripts/scaffolding.py [hunk 3]
```diff
@@ -137,14 +148,93 @@ def energy_intensity(df,df1,distance,col1,col2):
...
+    """
+
+    # Create a copy of the dura dataframe
+    dura = dura.copy()
+
+    # Create a replaced mode column in dura same as mode
+    dura[repm] = dura['mode']
+
+    # Pair dura with mode
+    dic_dura__trip = dict(zip(dura[repm],dura['C($/PMT)']))
```
Comment (shankari) [inline viz_scripts/scaffolding.py:208]: not sure why you are using `$/PMT` here...

File: viz_scripts/scaffolding.py [hunk 4]
```diff
@@ -207,3 +297,67 @@ def CO2_impact_lb(df,distance,col1,col2):
...
+        repm - feature name in df of feature with replaced mode
+        mode - feature name in df of feature with confirmed mode
+        
+    Returns:
+        data with appended time impact feature for each trip in $$$ (float)
+    """
+
+    data[mode+'_dura'] = data[dist] * data['dura__trip_mode']
+    data[repm+'_dura'] = data[dist] * data['dura__trip_repm']
+    data['Cost_Impact($)'] = round((data[mode+'_dura'] - data[repm+'_dura']),3)
```
Comment (shankari) [inline viz_scripts/scaffolding.py:339]: Ditto, why is this the cost impact?

File: viz_scripts/unit_tests.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+            'ei_mode': [1,2,3,1],
+            'ei_repm': [1,1,2,3],
+            'ei_trip_mode': [7,8,9,7],
+            'ei_trip_repm': [7,7,8,9],
+            'Mode_confirm_fuel': ['gasoline','diesel','electric','gasoline'],
+            'Replaced_mode_fuel': ['gasoline','gasoline','diesel','electric']
+        })
+
+
+    def test_process(self):
```
Comment (shankari) [inline viz_scripts/unit_tests.py:114]: I like the name of `test_process` versus `test_function`...

**Review by shankari — CHANGES_REQUESTED** [review]
Mostly documentation changes, but also some real code changes.

File: viz_scripts/scaffolding.py [hunk 6]
```diff
@@ -105,7 +105,7 @@ def data_quality_check(expanded_ct):
 
 def unit_conversions(df):
     df['distance_miles']= df["distance"]*0.00062 #meters to miles
-    df['duration'] = df['end_ts'] - df['start_ts']
+    df['duration_h'] = (df['end_ts'] - df['start_ts']) / 60 / 60 #seconds to hours
```
Comment (shankari) [inline viz_scripts/scaffolding.py]: @styoung3 why do you need to have the duration in hours? We are using speeds from the dataset directly - it's not like we need to convert use an external mph value.

Comment (styoung3) [reply viz_scripts/scaffolding.py]: I calculate average speed using (distance / duration) for each trip. Then I aggregate by mode and calculate the mean. MPH was easier for me to understand.

File: viz_scripts/unit_tests.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        a_dict = dict(zipped)
+        self.assertEqual(expect, a_dict,
+                            'Dict malfunction')
+    
+        expect = pd.Series(
+            [12, 12, 5, 2]
+        )
+        a_dict = dict(zip(self.constants['mode'], self.constants['vals']))
+        output = self.data['repm'].map(a_dict)
+        self.assertTrue(expect.equals(output),
```
[the remark this answers is not in the document]
Comment (styoung3) [reply viz_scripts/unit_tests.py]: Good idea. Will explore when I have time.

File: viz_scripts/scaffolding.py (hunk 3 above)
Comment (styoung3) [reply viz_scripts/scaffolding.py]: Right, that needed to be fixed.

File: viz_scripts/unit_tests.py (hunk 5 above)
Comment (styoung3) [reply viz_scripts/unit_tests.py]: I used test_process to go through the function step by step. Then I used test_function to actually pass parameters to the function and evaluate the overall output.

**Pull request closed without merging**